subprojects/composite-builds/src/main/java/org/gradle/composite/internal/DefaultRootBuildState.java

```diff
@@ -63,15 +65,20 @@ class DefaultRootBuildState extends AbstractCompositeParticipantBuildState imple
         this.listenerManager = listenerManager;
 
         BuildScopeServices buildScopeServices = getBuildServices();
-        BuildLifecycleController buildLifecycleController = getBuildController();
-        ExceptionAnalyser exceptionAnalyser = buildScopeServices.get(ExceptionAnalyser.class);
-        BuildOperationRunner buildOperationRunner = buildScopeServices.get(BuildOperationRunner.class);
-        BuildStateRegistry buildStateRegistry = buildScopeServices.get(BuildStateRegistry.class);
-        BuildTreeLifecycleControllerFactory buildTreeLifecycleControllerFactory = buildScopeServices.get(BuildTreeLifecycleControllerFactory.class);
-        BuildTreeWorkExecutor workExecutor = new BuildOperationFiringBuildTreeWorkExecutor(new DefaultBuildTreeWorkExecutor(), buildOperationRunner);
-        BuildTreeFinishExecutor finishExecutor = new OperationFiringBuildTreeFinishExecutor(buildOperationRunner,
-            new DefaultBuildTreeFinishExecutor(buildStateRegistry, exceptionAnalyser, buildLifecycleController));
-        this.buildTreeLifecycleController = buildTreeLifecycleControllerFactory.createRootBuildController(buildLifecycleController, workExecutor, finishExecutor);
+        try {
+            BuildLifecycleController buildLifecycleController = getBuildController();
+            ExceptionAnalyser exceptionAnalyser = buildScopeServices.get(ExceptionAnalyser.class);
+            BuildOperationRunner buildOperationRunner = buildScopeServices.get(BuildOperationRunner.class);
+            BuildStateRegistry buildStateRegistry = buildScopeServices.get(BuildStateRegistry.class);
+            BuildTreeLifecycleControllerFactory buildTreeLifecycleControllerFactory = buildScopeServices.get(BuildTreeLifecycleControllerFactory.class);
+            BuildTreeWorkExecutor workExecutor = new BuildOperationFiringBuildTreeWorkExecutor(new DefaultBuildTreeWorkExecutor(), buildOperationRunner);
+            BuildTreeFinishExecutor finishExecutor = new OperationFiringBuildTreeFinishExecutor(buildOperationRunner,
+                new DefaultBuildTreeFinishExecutor(buildStateRegistry, exceptionAnalyser, buildLifecycleController));
+            this.buildTreeLifecycleController = buildTreeLifecycleControllerFactory.createRootBuildController(buildLifecycleController, workExecutor, finishExecutor);
+        } catch (Throwable t) {
+            CompositeStoppable.stoppable().addFailure(t).add(buildScopeServices).stop();
+            throw UncheckedException.throwAsUncheckedException(t);
+        }
```

Comment:
🤔 IIUC, this is a typical use case where we are in one scope, and we want to execute code in a more specific scope with additional services. Feels like `ServiceRegistry` should understand this transition and offer some API to capture it. Then we wouldn't need this brittle-looking `try-catch` boilerplate.

Reply:
I don't think that's the problem. The complexity here is because these objects 1) create the build scoped services in their constructor and then 2) go on to do some more work in the constructor that can fail. This cleanup code would be here regardless of whether a service registry is involved or not.